repo/fsrepo/datastores.go

```diff
@@ -0,0 +1,287 @@
+}
+
+func MountDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	var res mountDatastoreConfig
+	mounts, ok := params["mounts"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("'mounts' field is missing or not an array")
+	}
+	for _, iface := range mounts {
+		cfg, ok := iface.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("expected map for mountpoint")
+		}
+
+		child, err := AnyDatastoreConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		prefix, found := cfg["mountpoint"]
+		if !found {
+			return nil, fmt.Errorf("no 'mountpoint' on mount")
+		}
+
+		res.mounts = append(res.mounts, premount{
+			ds:     child,
+			prefix: ds.NewKey(prefix.(string)),
+		})
+	}
+
+	return &res, nil
+}
+
+func (c *mountDatastoreConfig) DiskId() string {
+	buf := new(bytes.Buffer)
+	for _, m := range c.mounts {
+		fmt.Fprintf(buf, "%s:{%s};", m.prefix.String(), m.ds.DiskId())
+	}
+	return buf.String()
+}
+
+func (c *mountDatastoreConfig) Create(path string) (repo.Datastore, error) {
+	mounts := make([]mount.Mount, len(c.mounts))
+	for i, m := range c.mounts {
+		ds, err := m.ds.Create(path)
+		if err != nil {
+			return nil, err
+		}
+		mounts[i].Datastore = ds
+		mounts[i].Prefix = m.prefix
+	}
+	return mount.New(mounts), nil
+}
+
+type flatfsDatastoreConfig struct {
+	path      string
+	shardFun  *flatfs.ShardIdV1
+	syncField bool
+}
+
+func FlatfsDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	var c flatfsDatastoreConfig
+	var ok bool
+	var err error
+
+	c.path, ok = params["path"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'path' field is missing or not boolean")
+	}
+
+	sshardFun, ok := params["shardFunc"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'shardFunc' field is missing or not a string")
+	}
+	c.shardFun, err = flatfs.ParseShardFunc(sshardFun)
+	if err != nil {
+		return nil, err
+	}
+
+	c.syncField, ok = params["sync"].(bool)
+	if !ok {
+		return nil, fmt.Errorf("'sync' field is missing or not boolean")
+	}
+	return &c, nil
+}
+
+func (c *flatfsDatastoreConfig) DiskId() string {
+	return fmt.Sprintf("flatfs;%s;%s", c.path, c.shardFun.String())
+}
+
+func (c *flatfsDatastoreConfig) Create(path string) (repo.Datastore, error) {
+	p := c.path
+	if !filepath.IsAbs(p) {
+		p = filepath.Join(path, p)
+	}
+
+	return flatfs.CreateOrOpen(p, c.shardFun, c.syncField)
+}
+
+type leveldsDatastoreConfig struct {
+	path        string
+	compression ldbopts.Compression
+}
+
+func LeveldsDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	var c leveldsDatastoreConfig
+	var ok bool
+
+	c.path, ok = params["path"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'path' field is missing or not string")
+	}
+
+	switch params["compression"].(string) {
+	case "none":
+		c.compression = ldbopts.NoCompression
+	case "snappy":
+		c.compression = ldbopts.SnappyCompression
+	case "":
+		fallthrough
+	default:
+		c.compression = ldbopts.DefaultCompression
```

Comment:
This looks like it ignores garbage values (e.g., "snapy" will silently ignored). if so, this should probably return an error.